### PR TITLE
feat(Android): Implement ionic-file and ionic-content urls

### LINKED ISSUE #183: Images without file extensions are loaded as text/html content type
Version of plugin: 2.1.4
Android version 9
Android chrome version 69.0.3497.100
-------

When loading images that do not have a file extension, they are served as text/html which breaks the browser (`img` tags, etc.) This causes images to be displayed as "missing" or blank, and the placeholder broken image link icon appears where the image should be.

This used to work fine when images were served from file://, but no longer with the 2.0 update using the http://localhost:8080/_file_, approach. I'm not sure if this is a problem with the browser or the built-in server.

Example URL that worked with 1.2.1/x: `file:///data/user/0/com.maritlabs.topdecked.mtg/cache/image-loader-cache/551799067`

But in 2.0 this breaks...

**For example, an URL such as:** `http://localhost:8080/data/user/0/com.maritlabs.topdecked.mtg/cache/image-loader-cache/551799067` (file is correctly named/addressed)

Displays a broken link with this plugin installed, it displays fine

**However the following URL displays fine** when the image extension is added to the url (assuming the actual file on disk is renamed as well): 
`http://localhost:8080/data/user/0/com.maritlabs.topdecked.mtg/cache/image-loader-cache/551799067.png` (file extension has been added to both file and address)

This seems like a problem with file-type detection on the server/browser, but I'm not actually sure how this is *supposed* to behave.

### LINKED ISSUE #204: Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
I am using contacts plugin and unable to load URL which looks like this:
`content://com.android.contacts/contacts/8/photo`

`this.webview.convertFileSrc(contacts[i].photos[0].value);`

```
$ ionic info
✔ Gathering environment info - done!

Ionic:

   ionic (Ionic CLI)             : 4.2.1 (/usr/local/lib/node_modules/ionic)
   Ionic Framework               : @ionic/angular 4.0.0-beta.13
   @angular-devkit/build-angular : 0.8.6
   @angular-devkit/schematics    : 0.8.6
   @angular/cli                  : 6.2.6
   @ionic/angular-toolkit        : 1.0.0

Cordova:

   cordova (Cordova CLI) : 8.0.0
   Cordova Platforms     : android 7.0.0
   Cordova Plugins       : cordova-plugin-ionic-keyboard 2.1.3, cordova-plugin-ionic-webview 2.2.0, (and 10 other plugins)

System:

   Android SDK Tools : 26.1.1 (/Users/raza/Library/Android/sdk)
   ios-deploy        : 1.9.4
   ios-sim           : 7.0.0
   NodeJS            : v8.11.3 (/usr/local/bin/node)
   npm               : 6.3.0
   OS                : macOS
   Xcode             : Xcode 10.0 Build version 10A255
```

### PR DESCRIPTION
**Pull request by jcesarmobile**
Implement ionic-file and ionic-content for Android

closes #204, closes #183

### CODE REVIEW

File: src/android/com/ionicframework/cordova/webview/WebViewLocalServer.java [hunk 1]
```diff
@@ -48,12 +48,10 @@
...
   private static String TAG = "WebViewAssetServer";
   private String basePath;
-  /**
-   * capacitorapp.net is reserved by the Ionic team for use in local capacitor apps.
-   */
-  public final static String knownUnusedAuthority = "capacitorapp.net";
   private final static String httpScheme = "http";
   private final static String httpsScheme = "https";
+  private final static String ionicFileScheme = "ionic-file";
+  private final static String ionicContentScheme = "ionic-content";
```
Comment (mlynch) [inline src/android/com/ionicframework/cordova/webview/WebViewLocalServer.java]: See comment above re: `ionic-content` and `ionic-file`

File: src/www/util.js [hunk 2]
```diff
@@ -5,17 +5,14 @@ var WebView = {
...
-    if (window.WEBVIEW_SERVER_URL.startsWith('ionic://')) {
-      return url.replace('file', 'ionic-asset');
+    if (url.startsWith('file://')) {
+      return url.replace('file', 'ionic-file');;
     }
-    url = url.substr(7); // len("file://") == 7
-    if (url.length === 0 || url[0] !== '/') { // ensure the new URL starts with /
-      url = '/' + url;
+    if (url.startsWith('content://')) {
+        return url.replace('content://', 'ionic-content:///');;
```
Comment (mlynch) [inline src/www/util.js]: Watch for hard coded strings here

File: src/android/com/ionicframework/cordova/webview/AndroidProtocolHandler.java [hunk 3]
```diff
@@ -82,6 +72,16 @@ public InputStream openFile(String filePath) throws IOException  {
     return new FileInputStream(localFile);
   }
 
+  public InputStream openContentUrl(Uri uri)  throws IOException {
+    InputStream stream = null;
+    try {
+      stream = context.getContentResolver().openInputStream(Uri.parse(uri.toString().replace("ionic-content:///", "content://")));
```
[the remark this answers is not in the document]
Comment (jcesarmobile) [reply src/android/com/ionicframework/cordova/webview/AndroidProtocolHandler.java]: Changed to `app-content` and `app-file`, but as it's "Ionic WebView" I think it makes sense to use ionic as prefix on the schemes.
This schemes are only understood by the WebView, and we provide the logic to convert the file/content urls to something the WebView understands.

Specially in Capacitor I think it makes sense to have them like `capacitor-*`

File: src/www/util.js (hunk 2 above)
Comment (jcesarmobile) [reply src/www/util.js]: Addressed on last commit

**Comment by Bramzor** [comment]
Just tried this pull request but was not able to load android content. Also seems that the version for ionic does not work. Are there any other prerequisites apart from
cordova plugin rm cordova-plugin-ionic-webview --force
cordova plugin add cordova-plugin-ionic-webview@latest
To make sure cordova-plugin-ionic-webview 2.3.1 is installed. And use the function in util.js? I also noticed that in the latest git, ionic-content was replaced by a VARIABLE of which I cannot find the definition in this repo. So I tried replacing content:// with ionic-content:/// but that didnt work.

**Comment by jcesarmobile** [comment]
This is unreleased, will be available on next release (3.0.0)
If you want to test this you can install from github using the repository url.
Also if you use angular you have to tell it to trust this url schemes, you can do it in different ways, one of them is using SafeReourceUrl and DomSanitizer

**Comment by Bramzor** [comment]
Currently for iOS I use 'http://localhost:8080/_file_' to get to local iOS content. Is there a same way to reach the android content:// which I can use before this PR is released?

**Comment by jcesarmobile** [comment]
Use window.Ionic.WebView.convertFileSrc on the urls, it should always return a valid url depending on the webview plugin version

**Comment by Bramzor** [comment]
> Use window.Ionic.WebView.convertFileSrc on the urls, it should always return a valid url depending on the webview plugin version

For iOS this didnt work as the plugin (cordova-plugin-contacts) does not return a file:// on iOS but an url like /var/mobile/Containers/Data/Application/... (so you might want to add this to convertFileSrc ;-) .So I was able to get this file by appending the url like I mentioned. Had a lot of issues trying to use that convertFileSrc function so in the end I gave up trying that approach. Dont remember if I pushed it all the way to notice that it just didnt work or that I just gave up.

**Pull request merged**